### ISSUE TITLE
Run as crontab

### ISSUE THREAD

**Issue by Wtower** [UtkarshGpta/bing-desktop-wallpaper-changer#12]
The script does not seem to be able to run as a user crontab:

```
$ crontab -e
> 57 * * * * /usr/bin/python /home/wtower/.local/share/bing-desktop-wallpaper-changer-master/main.py
```

The log does not show anything wrong:

```
$ tail /var/log/syslog
> Aug  8 12:57:01 rodope CRON[20519]: (wtower) CMD (/usr/bin/python /home/wtower/.local/share/bing-desktop-wallpaper-changer-master/main.py)
```

It runs beautifully standalone. Running Ubuntu 16.04.

Any ideas?

**Comment by utkarshgpta** [comment]
I am sorry, I've no idea about that!

**Comment by Sim4n6** [comment]
Hi, I would like to help. Can you please provide the output of the following commands :  
`more /etc/os-release`
`users` and `who -a`
`sudo service cron status`
`cat /etc/crontab`

This commands will help me diagnostic your problem :+1:

**Comment by Wtower** [comment]
Hi, sure.

``` bash
gkarak@rodope ~> more /etc/os-release
NAME="Ubuntu"
VERSION="16.04.1 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.1 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
UBUNTU_CODENAME=xenial
gkarak@rodope ~> users
gkarak
gkarak@rodope ~> who -a
           system boot  2016-08-08 19:20
           επίπεδο-εκτέλεσης 5 2016-08-08 19:21
LOGIN      tty1         2016-08-08 19:21              3923 id=tty1
gkarak   + tty7         2016-08-08 19:27  πα�       4276 (:0)
gkarak@rodope ~> cat /etc/crontab 
# /etc/crontab: system-wide crontab
# Unlike any other crontab you don't have to run the `crontab'
# command to install the new version when you edit this file
# and files in /etc/cron.d. These files also have username fields,
# that none of the other crontabs do.

SHELL=/bin/sh
PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin

# m h dom mon dow user  command
17 *    * * *   root    cd / && run-parts --report /etc/cron.hourly
25 6    * * *   root    test -x /usr/sbin/anacron || ( cd / && run-parts --report /etc/cron.daily )
47 6    * * 7   root    test -x /usr/sbin/anacron || ( cd / && run-parts --report /etc/cron.weekly )
52 6    1 * *   root    test -x /usr/sbin/anacron || ( cd / && run-parts --report /etc/cron.monthly )
#
gkarak@rodope ~> 
```

**Comment by Sim4n6** [comment]
Service status?

**Comment by Wtower** [comment]
```bash
gkarak@rodope ~> sudo service --status-all
 [ + ]  acpid
 [ - ]  alsa-utils
 [ - ]  anacron
 [ + ]  apparmor
 [ + ]  apport
 [ + ]  avahi-daemon
 [ + ]  binfmt-support
 [ - ]  bluetooth
 [ - ]  bootmisc.sh
 [ - ]  brltty
 [ + ]  cgmanager
 [ - ]  cgproxy
 [ - ]  checkfs.sh
 [ - ]  checkroot-bootclean.sh
 [ - ]  checkroot.sh
 [ + ]  console-setup
 [ + ]  cron
 [ - ]  cups
 [ + ]  cups-browsed
 [ + ]  dbus
 [ - ]  dns-clean
 [ + ]  grub-common
 [ - ]  hostname.sh
 [ - ]  hwclock.sh
 [ + ]  irqbalance
 [ - ]  kerneloops
 [ + ]  keyboard-setup
 [ - ]  killprocs
 [ + ]  kmod
 [ + ]  lightdm
 [ + ]  lm-sensors
 [ + ]  mdadm
 [ - ]  mdadm-waitidle
 [ + ]  mongodb
 [ - ]  mountall-bootclean.sh
 [ - ]  mountall.sh
 [ - ]  mountdevsubfs.sh
 [ - ]  mountkernfs.sh
 [ - ]  mountnfs-bootclean.sh
 [ - ]  mountnfs.sh
 [ + ]  mysql
 [ + ]  network-manager
 [ + ]  networking
 [ + ]  nmbd
 [ + ]  ondemand
 [ - ]  plymouth
 [ - ]  plymouth-log
 [ + ]  postfix
 [ - ]  pppd-dns
 [ + ]  procps
 [ + ]  rc.local
 [ + ]  resolvconf
 [ - ]  rsync
 [ + ]  rsyslog
 [ + ]  samba
 [ + ]  samba-ad-dc
 [ - ]  saned
 [ - ]  sendsigs
 [ + ]  smbd
 [ + ]  speech-dispatcher
 [ + ]  ssh
 [ - ]  thermald
 [ + ]  udev
 [ + ]  ufw
 [ - ]  umountfs
 [ - ]  umountnfs.sh
 [ - ]  umountroot
 [ - ]  unattended-upgrades
 [ + ]  urandom
 [ - ]  uuidd
 [ + ]  virtualbox
 [ + ]  whoopsie
 [ - ]  x11-common
gkarak@rodope ~> 
```

**Comment by Sim4n6** [comment]
I mean only cron service status please

**Comment by Wtower** [comment]
``` bash
gkarak@rodope ~> sudo service cron status
● cron.service - Regular background program processing daemon
   Loaded: loaded (/lib/systemd/system/cron.service; enabled; vendor preset: ena
   Active: active (running) since Δευ 2016-08-08 19:20:57 EEST; 2 days ago
     Docs: man:cron(8)
 Main PID: 2973 (cron)
   CGroup: /system.slice/cron.service
           └─2973 /usr/sbin/cron -f

Αύγ 10 18:17:01 rodope CRON[21146]: pam_unix(cron:session): session opened for user root by (uid=0)
Αύγ 10 18:17:01 rodope CRON[21147]: (root) CMD (   cd / && run-parts --report /etc/cron.hourly)
Αύγ 10 19:17:01 rodope CRON[23365]: pam_unix(cron:session): session opened for user root by (uid=0)
Αύγ 10 19:17:01 rodope CRON[23366]: (root) CMD (   cd / && run-parts --report /etc/cron.hourly)
Αύγ 10 20:17:01 rodope CRON[25055]: pam_unix(cron:session): session opened for user root by (uid=0)
Αύγ 10 20:17:01 rodope CRON[25056]: (root) CMD (   cd / && run-parts --report /etc/cron.hourly)
Αύγ 10 21:17:01 rodope CRON[26181]: pam_unix(cron:session): session opened for user root by (uid=0)
Αύγ 10 21:17:01 rodope CRON[26182]: (root) CMD (   cd / && run-parts --report /etc/cron.hourly)
Αύγ 10 22:17:01 rodope CRON[27226]: pam_unix(cron:session): session opened for user root by (uid=0)
Αύγ 10 22:17:01 rodope CRON[27227]: (root) CMD (   cd / && run-parts --report /etc/cron.hourly)
gkarak@rodope ~> 
```

**Comment by Sim4n6** [comment]
How did you manage to cron the script please?

**Comment by Wtower** [comment]
What do you mean? You can see the crontab entry in my initial post:

> ```
> $ crontab -e
> > 57 * * * * /usr/bin/python /home/wtower/.local/share/bing-desktop-wallpaper-changer-master/main.py
> ```

**Comment by OSryx** [comment]
Ah sorry

**Comment by Sim4n6** [comment]
Can you provide the output of : 
`crontab -l`
`locate cron.deny`
`locate cron.allow`

**Comment by Sim4n6** [comment]
found it :smile: ... a clue : there is a python shebang in the script so it need a chmod +x before cron it 

please let me know if it worked, so that we close this issue.

**Comment by Wtower** [comment]
The shebang and execute permission are only required to execute the script directly as:

```
/home/gkarak/.local/share/bing-desktop-wallpaper-changer-master/main.py
```

In the crontab I execute python, which in turn executes the provided script and does _not_ require the execute bit.

Naturally, I tried it and it made no difference.

Regarding the output of the last commands you asked, I am not quite sure where you are heading at. The output of `crontab -l` is the same as provided in my initial post, as I already said. There are no other `cron.deny` or `cron.allow` files.

**Comment by Sim4n6** [comment]
This is really challenging @wtower

**Comment by Sim4n6** [comment]
I am having the same problem but in another different way : segfault ! 

```
Aug 13 12:04:01 HP CRON[8300]: (alji) CMD (/usr/bin/python3 /usr/bin/bing-desktop-wallpaper-changer/main.py )
Aug 13 12:04:03 HP kernel: [ 4978.581865] python3[8301]: segfault at 0 ip 00007f6d7ab46db9 sp 00007ffe07ad1c70 error 4 in libgtk-3.so.0.1800.9[7f6d7a88c000+71a000]
Aug 13 12:04:03 HP-15 CRON[8299]: (CRON) info (No MTA installed, discarding output)
```

**Comment by Sim4n6** [comment]
the MTA is just because cron deamon is unable to contact me by mail .... but the segfault is a serious matter. @Wtower what does your _syslog_ say after the cron execution line ?

**Comment by truongan** [comment]
Well, I don't use cron so I may not be of much help here. But FYI, I have been successfully created a **_systemd.timer**_ (a cron equivalent for system running **_systemd**_ like Fedora, ubuntu 16.04+, etc.) So the problem may not reside in the application but rather the configuration.

**Comment by Sim4n6** [comment]
I succeded to receive by local mail the message of error : 

```
From: root@HP-15-Notebook-PC (Cron Daemon)
To: alji@HP-15-Notebook-PC
Subject: Cron <alji@hp-15-notebook-pc> /usr/bin/bing-desktop-wallpaper-changer/main.py
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit
X-Cron-Env: <SHELL=/bin/sh>
X-Cron-Env: <PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin>
X-Cron-Env: <HOME=/home/alji>
X-Cron-Env: <LOGNAME=alji>
Message-Id: <20160813221202.06DF8A01ABF@HP-15-Notebook-PC>
Date: Sat, 13 Aug 2016 23:12:01 +0100 (WEST)

Failed to connect to Mir: Failed to connect to server socket: No such file or directory
Unable to init server: Could not connect: Connection refused
Failed to connect to Mir: Failed to connect to server socket: No such file or directory
Unable to init server: Could not connect: Connection refused

(main.py:6858): Gtk-CRITICAL **: gtk_settings_get_for_screen: assertion 'GDK_IS_SCREEN (screen)' failed
/usr/lib/python2.7/dist-packages/gi/overrides/__init__.py:326: Warning: g_object_get_qdata: assertion 'G_IS_OBJECT (object)' failed
  return super_init_func(self, **new_kwargs)
Segmentation fault
```

**Comment by truongan** [comment]
@OSryx I noticed that email is sent from root. Are you sure that cron job is setup for the right user?

**Comment by Sim4n6** [comment]
I edited /etc/crontab with user alji. Pretty sure that is not the problem.
The thing is cron environment is different from bash env... Still searching
to figure this out

**Comment by Wtower** [comment]
First, regarding @OSryx 's comment of the crontab user. Nice observation from @truongan. Of course the crontab user matters. The command `crontab -e` that I am using refers to my user's personal crontab, not the system-wide crontab. It makes no sense to add the script system-wide because it would run as root and would attempt to only change the wallpaper of root. editing the system-wide crontab file as whatever user still does not make any difference as the crontab is still root's.

So, to clear things up, my initial issue report concerns the **current user's crontab** and not a system-wide cron job. I am not interested in that, a lot of changes should happen to facilitate this otherwise.

Second, fedora's systemd timer has got absolutely nothing to do with cron, so I am reluctant to connect the two issues together, other than that they may share the same cause.

Third, regarding the segmentation fault, you have to make sure that lib gtk and all relevant dependencies are up to date. I am not getting this, as I initially posted.

**Comment by truongan** [comment]
Well, since I user systemd.timer, not crontab I cannot reproduce this problem. But according to my experience with systemd.timer I noticed that the application can run If and only If the user is currently logged in to a gnome session. If the user login from virtual console or currently in lock screen, it won't run.
Can you double check that the current user were currently in a gome session when the cron job fired.

**Comment by Wtower** [comment]
Yes, I would agree. The cron's user is logged in to a full desktop session, but the cron job runs in a tty and probably that's causing the issue.

**Comment by truongan** [comment]
Well, the python-gi library require by the application can only work with an active GDK_SCREEN to run. I can imagine invoke the gsettings command to set wallpaper as an alternative to run in tty console. I will try that in my fork to see how it go from there

**Comment by choomchichi** [comment]
Hi guys
This worked for me..

`* * * * * DISPLAY=:0 /full/path/to/main.py`

**Comment by Wtower** [comment]
@choomchichi Apparently this does not seem to work for me:

```
56 * * * * DISPLAY=:0 /usr/bin/python /home/wtower/.local/share/bing-desktop-wallpaper-changer-master/main.py
```

With no log entry. What distro you are using?

**Comment by choomchichi** [comment]
@Wtower 
Whats the output of this?
`echo $DISPLAY`

**Comment by Wtower** [comment]
It seems correct:

```
wtower@rodope ~> echo $DISPLAY 
:0
```

Of course the file gets downloaded, it just doesn't change.

**Comment by choomchichi** [comment]
let's check again..
If it runs correctly standalone and the log doesn't show anything helpful;
That seems stupid, but try this with execute permission:
`56 * * * * DISPLAY=:0 /home/wtower/.local/share/bing-desktop-wallpaper-changer-master/main.py`
and also try this:
`57 * * * * DISPLAY=:0 gedit`
...

**Comment by choomchichi** [comment]
@Wtower 
Forget my previous comments, and do this..
Edit bing-wallpapers.sh:

`#!/bin/sh`

`# export DBUS_SESSION_BUS_ADDRESS environment variable`
`PID=$(pgrep gnome-session)`
`export DBUS_SESSION_BUS_ADDRESS=$(grep -z DBUS_SESSION_BUS_ADDRESS /proc/$PID/environ|cut -d= -f2-)`

`/usr/bin/python /home/wtower/.local/share/bing-desktop-wallpaper-changer-master/main.py`

And make sure about execute permission.

Then:
`crontab -e`
`* * * * * DISPLAY=:0 /full/path/to/bing-wallpapers.sh`

I hope this solves your problem...

**Comment by Wtower** [comment]
Thanks for your reply. It doesn't seem to work, this is the relevant log entry:

```
Nov 21 11:09:10 rodope dbus[1060]: [system] Activating via systemd: service name='org.freedesktop.hostname1' unit='dbus-org.freedesktop.hostname1.service'
Nov 21 11:09:10 rodope systemd[1]: Starting Hostname Service...
Nov 21 11:09:10 rodope dbus[1060]: [system] Successfully activated service 'org.freedesktop.hostname1'
Nov 21 11:09:10 rodope systemd[1]: Started Hostname Service.
Nov 21 11:09:10 rodope org.gtk.vfs.Daemon[2413]: ** (process:10824): WARNING **: Couldn't create directory monitor on smb://x-gnome-default-workgroup/. Error: Operation not supported by backend
```

**Comment by choomchichi** [comment]
Run bing-wallpapers.sh standalone, and see the log entry..

**Comment by Wtower** [comment]
Good thought, I had a mistype in the file. It now works. Excellent advice @choomchichi .

I'd recommend that you fork the project and contribute back the file and instructions.

**Comment by choomchichi** [comment]
@Wtower
I'm glad..
I didn't fork anything before, but ok, I'll try that!!
:)

**Comment by Wtower** [comment]
@choomchichi if you got any questions ask them here already. It would be nice that you contribute this back to the repository.

**Comment by choomchichi** [comment]
I think it's better to append this tutorial to  README.md;
can anyone help me?

**Comment by Wtower** [comment]
You can also include the `bing-wallpapers.sh` file too. In your forked repository, edit the README file and add a new one and then create a pull request.

**Comment by choomchichi** [comment]
@Wtower
done!

**Comment by Wtower** [comment]
Nice!

Issue closed by merge request #46

**Comment by lrumancik** [comment]
@Wtower would you mind sharing the mistype you had in your file? I am also working on getting the bing desktop changer working via cron and am having the same exact issue. When I copy and paste the provided lines to my script, I get an error with grep. What is the specified delimiter supposed to be for the cut command?

I get these errors:
grep: /proc/959: Is a directory
grep: 4874/environ: No such file or directory

Thanks!

**Comment by Wtower** [comment]
@lrumancik I'm afraid I don't remember at all after two years, and I no longer have the particular installation (got formatted and haven't kept these scripts). I would suggest you to open a new issue altogether.

**Comment by lrumancik** [comment]
@Wtower okay, will do! Thanks for your response.

**Comment by lrumancik** [comment]
For anyone who may be having the same issue, I solved the above issue. I am running Ubuntu 18.04, and my pgrep gnome-session returns two PIDs to the PID variable which then messed up the path expansion of /proc/$PID/environ. One of the PIDs had my username as the uid while the other one had gdm for the uid. To get my script to run, I had to add --euid leah (my username) to the pgrep command to get the PID for my gnome-session. 

To summarize, I added the following to my cron script:

PID=$(pgrep --euid my_username gnome-session)
export DBUS_SESSION_BUS_ADDRESS=$(grep -z DBUS_SESSION_BUS_ADDRESS /proc/$PID/environ|cut -d= -f2-)

and added the DISPLAY=:0 to the cron entry using crontab -e
1 * * * * DISPLAY=:0 /full/path/to/bing-wallpapers.sh
